fuzz/fuzz_targets/refcounted_model.rs

```diff
@@ -7,125 +7,77 @@
+				compression: config.compression.into(),
+				btree_index: config.btree_index,
+				..parity_db::ColumnOptions::default()
+			}],
+			sync_wal: true,
+			sync_data: true,
+			stats: false,
+			salt: None,
+			compression_threshold: HashMap::new(),
+		}
+	}
 
-fuzz_target!(|entry: (Config, Vec<Action>)| {
-	let (config, actions) = entry;
-	let dir = tempdir().unwrap();
-	let options = parity_db::Options {
-		path: dir.path().to_owned(),
-		columns: vec![parity_db::ColumnOptions {
-			ref_counted: true,
-			compression: match config.compression {
-				CompressionType::NoCompression => parity_db::CompressionType::NoCompression,
-				CompressionType::Snappy => parity_db::CompressionType::Snappy,
-				CompressionType::Lz4 => parity_db::CompressionType::Lz4,
+	fn apply_operation_on_model(operation: &Operation, model: &mut Self::Model) {
+		match *operation {
+			Operation::Set(k) => {
+				*model.entry(k).or_default() += 1;
 			},
-			btree_index: config.btree_index,
-			..parity_db::ColumnOptions::default()
-		}],
-		sync_wal: true,
-		sync_data: true,
-		stats: false,
-		salt: None,
-	};
-	let mut db = parity_db::Db::open_or_create(&options).unwrap();
-	let mut model = BTreeMap::<u8, (usize, u8)>::default();
-	for action in actions {
-		// We apply the action on both the database and the model
-		match action {
-			Action::Transaction(operations) => {
-				db.commit_changes(operations.iter().copied().map(|op| {
-					(
-						0u8,
-						match op {
-							Operation::Set(k, v) => parity_db::Operation::Set(vec![k], vec![v]),
-							Operation::Dereference(k) => parity_db::Operation::Dereference(vec![k]),
-							Operation::Reference(k) => parity_db::Operation::Reference(vec![k]),
-						},
-					)
-				}))
-				.unwrap();
-				for op in operations {
-					match op {
-						Operation::Set(k, v) => {
-							model.insert(k, (1, v));
-						},
-						Operation::Dereference(k) => {
-							if let Entry::Occupied(mut e) = model.entry(k) {
-								let (mut counter, _) = e.get_mut();
-								counter -= 1;
-								if counter == 0 {
-									e.remove_entry();
-								}
-							}
-						},
-						Operation::Reference(k) =>
-							if let Some((counter, _)) = model.get_mut(&k) {
-								*counter += 1;
-							},
+			Operation::Dereference(k) =>
+				if let Entry::Occupied(mut e) = model.entry(k) {
+					let counter = e.get_mut();
+					*counter -= 1;
+					if *counter == 0 {
+						e.remove_entry();
 					}
-				}
-			},
-			Action::Restart =>
-				db = {
-					drop(db);
-					parity_db::Db::open_or_create(&options).unwrap()
+				},
+			Operation::Reference(k) =>
+				if let Entry::Occupied(mut e) = model.entry(k) {
+					*e.get_mut() += 1;
 				},
 		}
+	}
 
-		// We check the state
-		for (k, (_, v)) in &model {
-			assert_eq!(db.get(0, &[*k]).unwrap(), Some(vec![*v]));
+	fn map_operation(operation: &Operation) -> parity_db::Operation<Vec<u8>, Vec<u8>> {
+		match *operation {
+			Operation::Set(k) => parity_db::Operation::Set(vec![k], vec![k]),
```

Comment:
We could use different length of keys and values here, either with smallrng or directly with fuzzer input (with fuzzer input should only make sense for btree).
But that's probably already on your roadmap.

Reply:
Yes, definitely. It's on my roadmap. The tricky thing is that we want keys to get often reused between operations.

Reply:
in my trie fuzzer I just for half the delete use a random index in the model

Reply:
Sounds like a great idea! I was also thinking about generating from the fuzzer input a set of keys and then picking in this set when generating operations.